### PR TITLE
[WIP] feat: Static site builder.

### PR DESCRIPTION
**Pull request by amiller-gh**
Hi there 👋 Took a stab at addressing #9. __This is still a work in progress__ – but I wanted to get something up for early feedback. The feature still needs tests, and there are likely corner cases I'm missing, but it doesn't break any of the existing test suite as written.

There were some larger architecture things that I considered playing with in this PR... for example:

>It would be nice to introduce the concept of a `Builder` interface, that accepts a Vapid `Project` object as the first argument to its constructor. All builders (like `VapidServer` and `VapidBuilder` from this PR) would expose a `start(project: Project): Promise<Result>` and `stop(): Promise<Result>` interface for the CLI to call. This un-tangles Vapid Project configuration option / env variable ingestion from build pipeline details. 
>
>Then, because of this clear separation of concerns, we could break some of these internal infra pieces into smaller, logical packages with their own well-defined architecture docs and API footprints (ex: like we do over in [`@css-blocks/core`](https://github.com/linkedin/css-blocks/tree/master/packages/%40css-blocks/core#packages)) For Vapid, these would be things like `/Project`, `/Builder`, `/TemplateRenderer`, etc...

However! That all began to feel a little intrusive and out of scope 🙂 So, I tried to optimize for reduced diff size – happy to open an RFC to bounce around some architecture ideas if you're interested!

Anyway, this PR does the following:

 - Add `vapid build <src> [<dist>]` cli command.
 - Break out core `/lib/Vapid.js` file for basic project, env, and config ingestion.
 - `lib/VapidServer.js` extends core `Vapid` class for dev server.
 - `lib/VapidBuilder.js` extends core `Vapid` class for static build.
 - Static builder generates static site under `/dist` by default.
 - Lots of Vapid specific path logic abstracted out to `lib/paths.js` util file.
 - Added `/dist` to template .gitignore.
 - Pulled template rendering functionality to `lib/renderer.js` util for sharing.
 - Abstracted some middleware business logic to path, renderer and favicon utils.

You can then serve the static project that is built into `/dist` with a very simple static file server:

```javascript
const express = require('express');
const app = express();
app.use(express.static('dist', { extensions: ['html'] }));
app.listen(3000, () => console.log(`Example app listening on port ${3000}!`));
```

Thank you for making this amazing project! I'm so excited to see where this goes, and help out where I can 👍

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -3,7 +3,10 @@
   "plugins": ["jest"],
   "rules": {
     "no-underscore-dangle": "off",
-    "no-use-before-define": "off"
+    "no-use-before-define": "off",
+    "no-continue": "off",
```
Comment (amiller-gh) [inline .eslintrc]: Happy to change these back and refactor if you feel strongly about these linting rules.

File: lib/middleware/favicon.js [hunk 2]
```diff
@@ -1,47 +1,52 @@
...
       ctx.type = 'image/x-icon';
-      ctx.body = icon;
+      ctx.body = filePath ? fs.readFileSync(filePath) : '';
     }
 
     return true;
   };
 };
+
+module.exports.findFavicon = findFavicon;
```
Comment (amiller-gh) [inline lib/middleware/favicon.js]: This feels weird, but I couldn't come up with a better way to share this business logic between `VapidServer` and `VapidBuilder` without a larger utils refactor.

**Comment by srobbin** [comment]
@amiller-gh Thanks so much for submitting this, and please accept my apologies for the slow reply to your comment in the issue. I'll take a look at this right after the holiday weekend. Cheers.

**Comment by amiller-gh** [comment]
No worries! Happy Turkey Day 🙂

**Comment by iKasch** [comment]
Hey Guys, i just want to chime in real quick while closely watching where this PR goes. I really love the idea of a build feature and actually have a use case for it right now. 
So i played around with the PR and it could really shape to be exactly what i was looking for. :)

Just to give you another use case, i have a client who uses Adobe Experience Manager. AEM only allows static content, you have to copy the html in AEM and make the assets accessible. So right now i am building the static site and maintaining the content inside the code.
Vapid as is could really help me maintaining the content but for AEM i need an export of some sort. 

Two other things i am personally missing (but might be more of an issue for the future if this PR grows), it would be amazing to configure export layouts like fractal has preview layouts: https://fractal.build/guide/components/preview-layouts.html#creating-a-preview-layout

and i think a build button inside the dashboard would be great :)

I don't want to rush this PR or build up any pressure or anything, i just really like what you guys did and can't wait to see where everything goes!

**Comment by srobbin** [comment]
@amiller-gh, @iKasch Thanks for the examples—they're super helpful.

How would you feel about expanding the scope of the feature slightly, and us creating a `vapid export` command with options to generate a 1) static site or 2) directory of templates w/ a JSON dump of the database? The common thread is that people often need a byproduct of the website for hosting, archives, or moving on to another CMS.

**Comment by amiller-gh** [comment]
We should also talk about static site export format 🙂There are two option – a fairly clean looking export format where you assume the host can strip `.html` file endings (like surge.sh):
```
index.html
about.html
/section-name
  - page1.html
  - page2.html
```

Or, a more dirty looking index-files-only format which will be supported by any static host with bare SPA style URLs:
```
index.html
/about
  - index.html
/section-name
  /page1
    - index.html
  /page2
    - index.html
```
It couldn't hurt to support both export formats.

I also think this is a good chance to add build time static asset filename hashing for easy CDN cache busting – that can probably be added in a separate PR though after exporter infra stabilizes a bit. 

I think Phase 1 of static builds can be CLI only. After we land that functionality, we can figure out if, how, and where it fits in to the dashboard in a separate PR.

@srobbin, happy to take a stab at making a more generalized base `Builder` design this weekend to enable all these export options if you're open to it 👍

**Comment by srobbin** [comment]
@amiller-gh My turn to apologize _again_: I'm sorry it's taken so long to reply. 

Re: export format, we might be able to infer .html vs index by crawling the templates, and looking at the existing links.

If you're still interested, I'd love for you to take a stab. Please let me know if there's anything I can do to help.

If we don't talk before then, have a Happy New year. Looking forward to collaborating with you in 2019.

**Comment by pbuzdin** [comment]
@amiller-gh hey!

How are you? ;)

**Comment by amiller-gh** [comment]
Ha, I'm good @pbuzdin! Haven't had time to work on this project any more since my initial branch 😬

**Comment by amiller-gh** [comment]
I'll finally be looping back to this in the next few weeks – I have a new project that'll require it! Excited to dive in again 🙂My plan is to get it to the point where we can land in master and I can get my projects off of a branch!
